File: Kapitel05/ollama-clients/python/demos/openai_compat.py
#!/usr/bin/env python3
"""
Demo: OpenAI SDK Kompatibilität

Zeigt, wie das OpenAI SDK mit Ollama verwendet werden kann.
Bestehender OpenAI-Code funktioniert mit minimalen Änderungen.
"""

from openai import OpenAI


# OpenAI Client auf Ollama zeigen
client = OpenAI(
    base_url="http://localhost:11434/v1",
    api_key="ollama"  # Wird von Ollama ignoriert
)


def basic_openai_chat():
    """Einfacher Chat mit OpenAI SDK."""
    print("=== OpenAI SDK mit Ollama ===\n")

    completion = client.chat.completions.create(
        model="llama3.2",
        messages=[
            {"role": "system", "content": "Du bist ein hilfreicher Assistent."},
            {"role": "user", "content": "Was ist REST API? Ein Satz."}
        ]
    )

    print("Antwort:", completion.choices[0].message.content)
    print("\nUsage:")
    if completion.usage:
        print(f"  - Prompt Tokens: {completion.usage.prompt_tokens}")
        print(f"  - Completion Tokens: {completion.usage.completion_tokens}")


def streaming_openai():
    """Streaming mit OpenAI SDK."""
    print("\n=== OpenAI Streaming mit Ollama ===\n")

    print("Frage: Was sind die SOLID-Prinzipien?\n")
    print("Antwort: ", end="", flush=True)

    stream = client.chat.completions.create(
        model="llama3.2",
        messages=[{
            "role": "user",
            "content": "Nenne die 5 SOLID-Prinzipien. Nur die Namen, eine Zeile pro Prinzip."
        }],
        stream=True
    )

    for chunk in stream:
        content = chunk.choices[0].delta.content or ""
        print(content, end="", flush=True)

    print("\n")


def function_calling_demo():
    """Function Calling (experimentell)."""
    print("\n=== Function Calling (experimentell) ===\n")

    try:
        response = client.chat.completions.create(
            model="llama3.2",
            messages=[{
                "role": "user",
                "content": "Wie ist das Wetter in Berlin?"
            }],
            tools=[{
                "type": "function",
                "function": {
                    "name": "get_weather",
                    "description": "Ruft das aktuelle Wetter für einen Ort ab",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "location": {
                                "type": "string",
                                "description": "Die Stadt, z.B. Berlin"
                            }
                        },
                        "required": ["location"]
                    }
                }
            }]
        )

        message = response.choices[0].message

        if message.tool_calls:
            print("Tool Call erkannt:")
            for call in message.tool_calls:
                print(f"  - Funktion: {call.function.name}")
                print(f"  - Argumente: {call.function.arguments}")
        else:
            print("Keine Tool Calls, normale Antwort:")
            print(message.content)

    except Exception as e:
        print(f"Function Calling nicht unterstützt: {e}")
        print("Verwenden Sie ein Modell wie llama3.1 oder qwen2.5 für Tool-Support.")


def embeddings_demo():
    """Embeddings mit OpenAI SDK."""
    print("\n=== Embeddings mit OpenAI SDK ===\n")

    response = client.embeddings.create(
        model="nomic-embed-text",
        input="Ollama macht lokale KI einfach."
    )

    embedding = response.data[0].embedding
    print(f"Embedding-Dimension: {len(embedding)}")
    print(f"Erste 5 Werte: [{', '.join(f'{v:.4f}' for v in embedding[:5])}...]")


def migration_guide():
    """Migrations-Anleitung ausgeben."""
    print("\n=== Migration von OpenAI zu Ollama ===\n")

    print("Vorher (OpenAI):")
    print('  client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])')
    print("  completion = client.chat.completions.create(")
    print('      model="gpt-4",')
    print("      messages=[...]")
    print("  )")

    print("\nNachher (Ollama):")
    print("  client = OpenAI(")
    print('      base_url="http://localhost:11434/v1",')
    print('      api_key="ollama"')
    print("  )")
    print("  completion = client.chat.completions.create(")
    print('      model="llama3.2",  # <- Nur Modellname ändern')
    print("      messages=[...]")
    print("  )")

    print("\n2 Zeilen Änderung - der Rest bleibt gleich!")


def main():
    basic_openai_chat()
    streaming_openai()
    function_calling_demo()
    embeddings_demo()
    migration_guide()


if __name__ == "__main__":
    main()

File: Kapitel05/ollama-clients/typescript/src/demos/openai-compat.ts
/**
 * Demo: OpenAI SDK Kompatibilität
 *
 * Zeigt, wie das OpenAI SDK mit Ollama verwendet werden kann.
 * Bestehender OpenAI-Code funktioniert mit minimalen Änderungen.
 */

import OpenAI from "openai";

// OpenAI Client auf Ollama zeigen
const client = new OpenAI({
  baseURL: "http://localhost:11434/v1",
  apiKey: "ollama", // Wird von Ollama ignoriert
});

async function basicOpenAIChat() {
  console.log("=== OpenAI SDK mit Ollama ===\n");

  const completion = await client.chat.completions.create({
    model: "llama3.2",
    messages: [
      { role: "system", content: "Du bist ein hilfreicher Assistent." },
      { role: "user", content: "Was ist REST API? Ein Satz." },
    ],
  });

  console.log("Antwort:", completion.choices[0].message.content);
  console.log("\nUsage:");
  console.log(`  - Prompt Tokens: ${completion.usage?.prompt_tokens}`);
  console.log(`  - Completion Tokens: ${completion.usage?.completion_tokens}`);
}

async function streamingOpenAI() {
  console.log("\n=== OpenAI Streaming mit Ollama ===\n");

  console.log("Frage: Was sind die SOLID-Prinzipien?\n");
  console.log("Antwort:");

  const stream = await client.chat.completions.create({
    model: "llama3.2",
    messages: [
      {
        role: "user",
        content: "Nenne die 5 SOLID-Prinzipien. Nur die Namen, eine Zeile pro Prinzip.",
      },
    ],
    stream: true,
  });

  for await (const chunk of stream) {
    const content = chunk.choices[0]?.delta?.content || "";
    process.stdout.write(content);
  }

  console.log("\n");
}

async function functionCallingDemo() {
  console.log("\n=== Function Calling (experimentell) ===\n");

  // Hinweis: Function Calling wird von einigen Modellen unterstützt
  try {
    const response = await client.chat.completions.create({
      model: "llama3.2",
      messages: [
        {
          role: "user",
          content: "Wie ist das Wetter in Berlin?",
        },
      ],
      tools: [
        {
          type: "function",
          function: {
            name: "get_weather",
            description: "Ruft das aktuelle Wetter für einen Ort ab",
            parameters: {
              type: "object",
              properties: {
                location: {
                  type: "string",
                  description: "Die Stadt, z.B. Berlin",
                },
              },
              required: ["location"],
            },
          },
        },
      ],
    });

    const message = response.choices[0].message;

    if (message.tool_calls) {
      console.log("Tool Call erkannt:");
      for (const call of message.tool_calls) {
        console.log(`  - Funktion: ${call.function.name}`);
        console.log(`  - Argumente: ${call.function.arguments}`);
      }
    } else {
      console.log("Keine Tool Calls, normale Antwort:");
      console.log(message.content);
    }
  } catch (error) {
    console.log("Function Calling nicht unterstützt von diesem Modell.");
    console.log("Verwenden Sie ein Modell wie llama3.1 oder qwen2.5 für Tool-Support.");
  }
}

async function embeddings() {
  console.log("\n=== Embeddings mit OpenAI SDK ===\n");

  const response = await client.embeddings.create({
    model: "nomic-embed-text",
    input: "Ollama macht lokale KI einfach.",
  });

  const embedding = response.data[0].embedding;
  console.log(`Embedding-Dimension: ${embedding.length}`);
  console.log(`Erste 5 Werte: [${embedding.slice(0, 5).map((n) => n.toFixed(4)).join(", ")}...]`);
}

async function migrateFromOpenAI() {
  console.log("\n=== Migration von OpenAI zu Ollama ===\n");

  console.log("Vorher (OpenAI):");
  console.log(`  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });`);
  console.log(`  const completion = await client.chat.completions.create({`);
  console.log(`    model: "gpt-4",`);
  console.log(`    messages: [...]`);
  console.log(`  });`);

  console.log("\nNachher (Ollama):");
  console.log(`  const client = new OpenAI({`);
  console.log(`    baseURL: "http://localhost:11434/v1",`);
  console.log(`    apiKey: "ollama"`);
  console.log(`  });`);
  console.log(`  const completion = await client.chat.completions.create({`);
  console.log(`    model: "llama3.2",  // <- Nur Modellname ändern`);
  console.log(`    messages: [...]`);
  console.log(`  });`);

  console.log("\n2 Zeilen Änderung - der Rest bleibt gleich!");
}

async function main() {
  await basicOpenAIChat();
  await streamingOpenAI();
  await functionCallingDemo();
  await embeddings();
  await migrateFromOpenAI();
}

main().catch(console.error);
